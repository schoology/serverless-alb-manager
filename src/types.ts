export interface AlbManagerOptions {
  vpcId: string;
  subnetIds: string[];
  tags?: Record<string, string>;
  certificateArn: string;
  domainName: string;
}
