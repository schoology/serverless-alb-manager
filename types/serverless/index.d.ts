// serverless type declaration fixes - to be removed once
// https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38315 is merged and released

// tslint:disable:no-any max-classes-per-file

declare module 'serverless/classes/Plugin' {
  const serverlessOriginal = require('serverless');
  declare interface Plugin {
    hooks: {
      [event: string]: () => Promise<any>;
    };
  }

  declare var Plugin: {
    prototype: Plugin;
    new (serverless: Serverless, options: Serverless.Options): Plugin;
  };
  export = Plugin;
}

declare module 'serverless/classes/Service' {
  declare class Service {
    public custom: Service.Custom;

    public provider: {
      compiledCloudFormationTemplate: {
        Resources: {
          [key: string]: any;
        };
        Outputs?: {
          [key: string]: any;
        };
      };

      name: string;
      stage: string;
      region: string;
      runtime?: string;
      timeout?: number;
      versionFunctions: boolean;

      naming: Record<string, () => string>;
    };
    constructor(serverless: Serverless, data: {});

    public load(rawOptions: {}): Promise<any>;
    public setFunctionNames(rawOptions: {}): void;

    public getServiceName(): string;
    public getAllFunctions(): string[];
    public getAllFunctionsNames(): string[];
    public getFunction(functionName: string): Serverless.FunctionDefinition;
    public getEventInFunction(eventName: string, functionName: string): Serverless.Event;
    public getAllEventsInFunction(functionName: string): Serverless.Event[];

    public mergeResourceArrays(): void;
    public validate(): Service;

    public update(data: {}): {};
  }

  export = Service;
}

declare module 'serverless/plugins/aws/provider/awsProvider' {
  declare class Aws {
    public naming: Record<string, () => string>;
    constructor(serverless: Serverless, options: Serverless.Options);
    public getProviderName(): string;
    public getRegion(): string;
    public getServerlessDeploymentBucketName(): string;
    public getStage(): string;
  }

  export = Aws;
}

declare module 'serverless' {
  const serverlessOriginal = require('serverless');

  declare namespace Serverless {
    export declare interface Event {
      alb?: {
        listenerArn?: string | object;
      };
    }
  }

  declare class Serverless extends serverlessOriginal {}

  export = Serverless;
}
