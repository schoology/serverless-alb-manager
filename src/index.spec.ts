import * as fs from 'fs';
import * as Serverless from 'serverless';
import Service = require('serverless/classes/Service');
import * as TypeMoq from 'typemoq';

const ServiceClass = require('serverless/lib/classes/Service'); // tslint:disable-line

import { ServerlessPluginAlbManager } from './index';
import { AlbManagerOptions } from './types';

/**
 * Mock a `custom.serverless-alb-manager` config from the Service mock
 */
const setConfig = (service: Service, config: object): void => {
  service.update({ custom: { ['serverless-alb-manager']: config } });
};

describe('ServerlessPluginAlbManager', () => {
  let serverlessMock: TypeMoq.IMock<Serverless>;
  let service: Service;
  let plugin: ServerlessPluginAlbManager;
  let providerObject: Service['provider'];
  let validConfig: AlbManagerOptions;

  beforeEach(() => {
    providerObject = {
      compiledCloudFormationTemplate: {
        Outputs: {},
        Resources: {},
      },
      name: 'dummy-provider-name',
      stage: 'dummy-provider-stage',
      region: 'dummy-provider-region',
      versionFunctions: true,
      naming: {
        getStackName: () => 'dummy-stack-name',
      },
    };

    validConfig = {
      vpcId: 'dummy-vpc-id',
      subnetIds: ['dummy-subnet-1', 'dummy-subnet-2'],
      certificateArn: 'dummy-certificate-arn',
      domainName: 'dummy.domain.example.com',
      tags: {
        dummyTag1: 'dummy-1',
        dummyTag2: 'dummy-2',
      },
    };

    serverlessMock = TypeMoq.Mock.ofType<Serverless>();

    service = new ServiceClass(serverlessMock.object, {
      provider: providerObject,
    });

    serverlessMock.setup(x => x.service).returns(() => service);
    serverlessMock.setup(x => x.getProvider(TypeMoq.It.isAnyString())).returns(() => providerObject);

    plugin = new ServerlessPluginAlbManager(serverlessMock.object);
  });

  describe('package:setupProviderConfiguration hook', () => {
    describe('with all expected config options', () => {
      beforeEach(() => {
        setConfig(service, validConfig);
      });

      it('generates a valid cloudformation template', async () => {
        await plugin.hooks['package:setupProviderConfiguration']();
        expect(service.provider.compiledCloudFormationTemplate).toMatchSnapshot();

        // Write out a JSON that will be validated during CI
        fs.writeFileSync(
          'template-snapshot.json',
          // tslint:disable-next-line:no-magic-numbers
          JSON.stringify(service.provider.compiledCloudFormationTemplate, undefined, 2),
        );
      });
    });

    describe('without tags', () => {
      beforeEach(() => {
        setConfig(service, {
          ...validConfig,
          tags: undefined,
        });
      });

      it('does not set any tags', async () => {
        await plugin.hooks['package:setupProviderConfiguration']();
        expect(service.provider.compiledCloudFormationTemplate.Resources.Alb.Properties.Tags).toBeUndefined();
      });
    });

    describe('with comma-separated subnet IDs', () => {
      beforeEach(() => {
        setConfig(service, {
          ...validConfig,
          subnetIds: 'dummy-subnet-3, dummy-subnet-4',
        });
      });

      it('splits the subnets into an array', async () => {
        await plugin.hooks['package:setupProviderConfiguration']();
        expect(service.provider.compiledCloudFormationTemplate.Resources.Alb.Properties.Subnets).toMatchObject([
          'dummy-subnet-3',
          'dummy-subnet-4',
        ]);
      });
    });

    describe('with no config options', () => {
      it('throws an error', () => {
        return expect(plugin.hooks['package:setupProviderConfiguration']()).rejects.toThrowError(
          'serverless-plugin-alb-manager: Please define ALB options in serverless.yml:custom.alb. See README.md for details.',
        );
      });
    });

    describe('with invalid config options', () => {
      beforeEach(() => {
        setConfig(service, {
          foo: 'bar',
        });
      });

      it('throws an error', () => {
        return expect(plugin.hooks['package:setupProviderConfiguration']()).rejects.toThrowError(
          'serverless-plugin-alb-manager: Invalid configuration. ValidationError: ' +
            '"vpcId" is required. ' +
            '"subnetIds" is required. ' +
            '"certificateArn" is required. ' +
            '"domainName" is required. ' +
            '"foo" is not allowed',
        );
      });
    });
  });

  describe('package:compileFunctions hook', () => {
    describe('with alb events that do not have listenerArn defined', () => {
      beforeEach(() => {
        service.update({
          functions: {
            dummyFunc1: {
              events: [{ alb: { conditions: { path: '/dummy11' } } }, { alb: { conditions: { path: '/dummy12' } } }],
            },
            dummyFunc2: {
              events: [{ alb: { conditions: { path: '/dummy21' } } }, { alb: { conditions: { path: '/dummy22' } } }],
            },
          },
        });
        service.setFunctionNames({});
      });

      it('sets the event listenerArns', async () => {
        await plugin.hooks['package:compileFunctions']();
        const expectedListener = {
          Ref: 'AlbHttpListener',
        };
        expect(service.getAllEventsInFunction('dummyFunc1')[0].alb.listenerArn).toMatchObject(expectedListener);
        expect(service.getAllEventsInFunction('dummyFunc1')[1].alb.listenerArn).toMatchObject(expectedListener);
        expect(service.getAllEventsInFunction('dummyFunc2')[0].alb.listenerArn).toMatchObject(expectedListener);
        expect(service.getAllEventsInFunction('dummyFunc2')[1].alb.listenerArn).toMatchObject(expectedListener);
      });
    });

    describe('with alb events that do have listenerArn defined', () => {
      beforeEach(() => {
        service.update({
          functions: {
            dummyFunc1: {
              events: [
                {
                  alb: {
                    conditions: { path: '/dummy11' },
                    listenerArn: 'dummyListenerArn',
                  },
                },
              ],
            },
          },
        });
        service.setFunctionNames({});
      });

      it('sets the event listenerArns', async () => {
        await plugin.hooks['package:compileFunctions']();
        expect(service.getAllEventsInFunction('dummyFunc1')[0].alb.listenerArn).toBe('dummyListenerArn');
      });
    });

    describe('with non-alb events', () => {
      const nonAlbEvents = [{ http: 'GET /dummyFunc1' }];
      beforeEach(() => {
        service.update({
          functions: {
            dummyFunc1: { events: nonAlbEvents },
          },
        });
        service.setFunctionNames({});
      });

      it('does not modify the events', async () => {
        await plugin.hooks['package:compileFunctions']();
        expect(service.getAllEventsInFunction('dummyFunc1')).toMatchObject(nonAlbEvents);
      });
    });
  });
});
