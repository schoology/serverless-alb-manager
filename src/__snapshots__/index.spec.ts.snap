// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServerlessPluginAlbManager package:setupProviderConfiguration hook with all expected config options generates a valid cloudformation template 1`] = `
Object {
  "Outputs": Object {},
  "Resources": Object {
    "Alb": Object {
      "Properties": Object {
        "IpAddressType": "ipv4",
        "Name": "dummy-stack-name-alb",
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Ref": "AlbSecurityGroup",
          },
        ],
        "Subnets": Array [
          "dummy-subnet-1",
          "dummy-subnet-2",
        ],
        "Tags": Array [
          Object {
            "Key": "dummyTag1",
            "Value": "dummy-1",
          },
          Object {
            "Key": "dummyTag2",
            "Value": "dummy-2",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "AlbDnsRecord": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "Alb",
              "DNSName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::GetAtt": Array [
              "Alb",
              "CanonicalHostedZoneID",
            ],
          },
        },
        "HostedZoneName": "example.com.",
        "Name": "dummy.domain.example.com",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "AlbHttpListener": Object {
      "Properties": Object {
        "Certificates": Array [
          Object {
            "CertificateArn": "dummy-certificate-arn",
          },
        ],
        "DefaultActions": Array [
          Object {
            "FixedResponseConfig": Object {
              "ContentType": "application/json",
              "MessageBody": "{ \\"error\\": \\"Forbidden\\" }",
              "StatusCode": 403,
            },
            "Order": 1,
            "Type": "fixed-response",
          },
        ],
        "LoadBalancerArn": Object {
          "Ref": "Alb",
        },
        "Port": 443,
        "Protocol": "HTTPS",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "AlbSecurityGroup": Object {
      "Properties": Object {
        "GroupDescription": "HTTPS for dummy-stack-name",
        "GroupName": "dummy-stack-name-https",
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "FromPort": "443",
            "IpProtocol": "tcp",
            "ToPort": "443",
          },
        ],
        "VpcId": "dummy-vpc-id",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
}
`;
